src/main/java/com/wasim/codefellowship/controllers/ApplicationUserController.java

```diff
@@ -45,18 +44,80 @@ public String getSignupPage() {
+        } else {
+            throw new ResourceNotFoundException("It's a 404");
+        }
+
         return "index.html";
     }
-    
+
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+    public class ResourceNotFoundException extends RuntimeException {
+        ResourceNotFoundException(String message) {
+            super(message);
+        }
+    }
+
+    @GetMapping ("/test")
+    public String getTestPage(Principal p, Model m) {
+        if(p != null) {
+            String username = p.getName();
+            ApplicationUser applicationUser = applicationUserRepository.findByUsername(username);
+
+            m.addAttribute("username", username);
+//            m.addAttribute("firstname", applicationUserUser.getFirstName());
+        }
+
+        return "test.html";
+    }
+
+    @GetMapping("/users/{id}")
+    public String getUserInfo(Model m, Principal p, @PathVariable Long id) {
+        if (p != null)  // not strictly required if your WebSecurityConfig is correct
+        {
+            String username = p.getName();
+            ApplicationUser applicationBrowsingUser = applicationUserRepository.findByUsername(username);
+
+            m.addAttribute("browsingUserUsername", username);
+//            m.addAttribute("browsingUserNickname", applicationBrowsingUser.getNickname());
+        }
+
+        ApplicationUser applicationUser = applicationUserRepository.findById(id).orElseThrow();
+        m.addAttribute("applicationUserUsername", applicationUser.getUsername());
+//        m.addAttribute("applicationUserNickname", applicationUser.getNickname());
+//        m.addAttribute("applicationUserId", applicationUser.getId());
+
+        m.addAttribute("testDate", LocalDateTime.now());
+
+        return "user-info.html";
+    }
+
+    @PutMapping("/users/{id}")
+    public RedirectView edituserInfo(Model m, Principal p, @PathVariable Long id, String username, RedirectAttributes redir) {
+
+        if(p != null && (p.getName().equals(username))) {
+            ApplicationUser applicationUser = applicationUserRepository.findById(id).orElseThrow();
+            applicationUser.setUsername(username);
+            applicationUserRepository.save(applicationUser);
+        } else {
+            redir.addFlashAttribute("errorMessage", "Cannot edit another user's page!");
+        }
+        return new RedirectView("users/" + id);
+    }
+
     @PostMapping("/signup")
-    public RedirectView postSignup(String username, String password, String firstName, String lastName, String bio, LocalDate dateOfBirth) {
-        ApplicationUser user = new ApplicationUser();
-        user.setUsername(username);
-        user.setFirstName(firstName);
-        user.setLastName(lastName);
-        user.setBio(bio);
-        user.setDateOfBirth(dateOfBirth);
+    public RedirectView postSignup(String username, String password, String firstName, String lastName, LocalDate dateOfBirth, String bio) {
```

Comment:
Make sure that these variable names exactly match the `id` attributes in your form fields (and remember that Java is case-sensitive!)